src/edid.c

```diff
@@ -95,14 +382,79 @@ static void decode_display_descriptor(const unsigned char *desc, struct monitor_
 	}
 }
 
-static void decode_descriptors(const unsigned char *edid, struct monitor_info *info)
+static void decode_detailed_timing(const unsigned char *timing, struct detailed_timing * detailed)
 {
-	for (int i = 0; i < 4; ++i) {
+	int bits;
+
+	enum stereo_type stereo[] = {
+		NO_STEREO, NO_STEREO, FIELD_RIGHT, FIELD_LEFT,
+		TWO_WAY_RIGHT_ON_EVEN, TWO_WAY_LEFT_ON_EVEN,
+		FOUR_WAY_INTERLEAVED, SIDE_BY_SIDE
+	};
+
+	detailed->pixel_clock = (timing[0x00] | timing[0x01] << 8) * 10000;
+	detailed->h_addr = timing[0x02] | ((timing[0x04] & 0xf0) << 4);
+	detailed->h_blank = timing[0x03] | ((timing[0x04] & 0x0f) << 8);
+	detailed->v_addr = timing[0x05] | ((timing[0x07] & 0xf0) << 4);
+	detailed->v_blank = timing[0x06] | ((timing[0x07] & 0x0f) << 8);
+	detailed->h_front_porch = timing[0x08] | get_bits(timing[0x0b], 6, 7) << 8;
+	detailed->h_sync = timing[0x09] | get_bits(timing[0x0b], 4, 5) << 8;
+	detailed->v_front_porch = get_bits(timing[0x0a], 4, 7) | get_bits(timing[0x0b], 2, 3) << 4;
+	detailed->v_sync = get_bits(timing[0x0a], 0, 3) | get_bits(timing[0x0b], 0, 1) << 4;
+	detailed->width_mm = timing[0x0c] | get_bits(timing[0x0e], 4, 7) << 8;
+	detailed->height_mm = timing[0x0d] | get_bits(timing[0x0e], 0, 3) << 8;
+	detailed->right_border = timing[0x0f];
+	detailed->top_border = timing[0x10];
+
+	detailed->interlaced = get_bit(timing[0x11], 7);
+
+	/* Stereo */
+	bits = get_bits(timing[0x11], 5, 6) << 1 | get_bit(timing[0x11], 0);
+	detailed->stereo = stereo[bits];
+
+	/* Sync */
+	bits = timing[0x11];
+
+	detailed->digital_sync = get_bit(bits, 4);
+	if (detailed->digital_sync) {
+		detailed->digital.composite = !get_bit(bits, 3);
+
+		if (detailed->digital.composite) {
+			detailed->digital.serrations = get_bit(bits, 2);
+			detailed->digital.negative_vsync = 0;
+		} else {
+			detailed->digital.serrations = 0;
+			detailed->digital.negative_vsync = !get_bit(bits, 2);
+		}
+
+		detailed->digital.negative_hsync = !get_bit(bits, 0);
+	} else {
+		detailed->analog.bipolar = get_bit(bits, 3);
+		detailed->analog.serrations = get_bit(bits, 2);
+		detailed->analog.sync_on_green = !get_bit(bits, 1);
+	}
+}
+
+static int decode_descriptors(const unsigned char *edid, struct monitor_info *info)
+{
+	int i;
+	int timing_idx;
+
+	timing_idx = 0;
+
+	for (i = 0; i < 4; ++i) {
 		int index = 0x36 + i * 18;
 
-		if (edid[index + 0] == 0x00 && edid[index + 1] == 0x00)
+		if (edid[index + 0] == 0x00 && edid[index + 1] == 0x00) {
```

Comment:
Awesome, you added curly braces around the if() even though it only had one line! I prefer having it in both the if and else, some kernel folks don't care or have strong opinions the other way around.